File: plugins/MRStashVRTagger/MRStashVRTagger.js
(function () {
  "use strict";

  const PluginApi = window.PluginApi;
  const React = window.React || PluginApi.React;
  const ReactDOM = window.ReactDOM || PluginApi.ReactDOM;
  const { useState, useEffect } = React;
  const ce = React.createElement;

  const LOG = (...a) => console.log("[MRStashVRTagger]", ...a);

  // ── GraphQL ────────────────────────────────────────────────────────────────

  async function gql(query, variables) {
    const res = await fetch("/graphql", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ query, variables }),
    });
    const json = await res.json();
    if (json.errors) throw new Error(json.errors[0].message);
    return json.data;
  }

  async function applyTagChange(sceneId, tagId, action) {
    // Fetch current tags, modify, send update
    const data = await gql(
      `query($id: ID!) { findScene(id: $id) { id tags { id } } }`,
      { id: sceneId }
    );
    const current = (data.findScene.tags || []).map(t => t.id);
    let next;
    if (action === "add") {
      if (current.includes(tagId)) return;
      next = [...current, tagId];
    } else {
      next = current.filter(id => id !== tagId);
    }
    await gql(
      `mutation($input: SceneUpdateInput!) {
        sceneUpdate(input: $input) { id }
      }`,
      { input: { id: sceneId, tag_ids: next } }
    );
  }

  async function runAuditTask() {
    await gql(
      `mutation RunPluginTask($plugin_id: ID!, $task_name: String!) {
        runPluginTask(plugin_id: $plugin_id, task_name: $task_name)
      }`,
      { plugin_id: "MRStashVRTagger", task_name: "Audit VR Tags" }
    );
  }

  // ── Audit fetch ────────────────────────────────────────────────────────────

  async function fetchAudit() {
    try {
      const res = await fetch(`/plugin/MRStashVRTagger/assets/audit.json?t=${Date.now()}`);
      if (!res.ok) return null;
      return await res.json();
    } catch (_) {
      return null;
    }
  }

  // ── Modal ──────────────────────────────────────────────────────────────────

  let _modalRoot = null;

  function openModal() {
    if (!_modalRoot) {
      _modalRoot = document.createElement("div");
      _modalRoot.id = "vra-modal-root";
      document.body.appendChild(_modalRoot);
    }
    ReactDOM.render(ce(AuditModal, { onClose: closeModal }), _modalRoot);
  }

  function closeModal() {
    if (_modalRoot && ReactDOM) ReactDOM.unmountComponentAtNode(_modalRoot);
  }

  function AuditModal({ onClose }) {
    const [audit, setAudit] = useState(null);
    const [loading, setLoading] = useState(true);
    const [filter, setFilter] = useState("all");
    const [busyIds, setBusyIds] = useState({});
    const [doneIds, setDoneIds] = useState({});
    const [running, setRunning] = useState(false);

    useEffect(() => {
      document.body.style.overflow = "hidden";
      fetchAudit().then(d => { setAudit(d); setLoading(false); });
      return () => { document.body.style.overflow = ""; };
    }, []);

    async function reRun() {
      setRunning(true);
      try { await runAuditTask(); } catch (_) {}
      // Poll for new audit.json (look for a newer generated_at)
      const oldStamp = audit && audit.generated_at;
      let attempts = 0;
      const iv = setInterval(async () => {
        attempts++;
        const fresh = await fetchAudit();
        if (fresh && fresh.generated_at && fresh.generated_at !== oldStamp) {
          clearInterval(iv);
          setAudit(fresh);
          setRunning(false);
          setDoneIds({});
        }
        if (attempts > 600) { clearInterval(iv); setRunning(false); }
      }, 1000);
    }

    async function applyOne(entry) {
      if (!audit || !audit.tag_id) return;
      setBusyIds(b => ({ ...b, [entry.id]: true }));
      try {
        await applyTagChange(entry.id, audit.tag_id, entry.action);
        setDoneIds(d => ({ ...d, [entry.id]: true }));
      } catch (e) {
        LOG("apply failed", e);
      } finally {
        setBusyIds(b => { const n = { ...b }; delete n[entry.id]; return n; });
      }
    }

    async function applyAll(list) {
      for (const entry of list) {
        if (doneIds[entry.id]) continue;
        await applyOne(entry);
      }
    }

    const all = audit ? [...audit.needs_add, ...audit.needs_remove] : [];
    const visible = filter === "add" ? (audit ? audit.needs_add : [])
                   : filter === "remove" ? (audit ? audit.needs_remove : [])
                   : all;

    return ce("div", {
      className: "vra-modal-overlay",
      onClick: e => { if (e.target === e.currentTarget) onClose(); }
    },
      ce("div", { className: "vra-modal-box" },
        ce("div", { className: "vra-modal-header" },
          ce("h2", null, "VR Tag Audit"),
          ce("button", { className: "vra-modal-close", onClick: onClose }, "✕")
        ),

        loading && ce("div", { className: "vra-status" }, "Loading audit results…"),

        !loading && !audit && ce("div", null,
          ce("div", { className: "vra-status" }, "No audit results yet. Run the audit task to generate one."),
          ce("div", { className: "vra-actions" },
            ce("button", { className: "vra-btn vra-btn-primary", onClick: reRun, disabled: running },
              running ? "Running…" : "Run Audit Now")
          )
        ),

        audit && ce(React.Fragment, null,
          ce("div", { className: "vra-subtitle" },
            `Scanned ${audit.total_scenes} scenes — threshold ${audit.threshold}:1 — tag "${audit.tag_name}"`
          ),
          !audit.tag_id && ce("div", { className: "vra-warn" },
            `Tag "${audit.tag_name}" doesn't exist yet. Create it in Stash before applying changes.`
          ),

          ce("div", { className: "vra-filter-bar" },
            ce("button", {
              className: `vra-pill ${filter==="all"?"active":""}`,
              onClick: () => setFilter("all")
            }, `All (${all.length})`),
            ce("button", {
              className: `vra-pill add ${filter==="add"?"active":""}`,
              onClick: () => setFilter("add")
            }, `Add VR (${audit.needs_add.length})`),
            ce("button", {
              className: `vra-pill remove ${filter==="remove"?"active":""}`,
              onClick: () => setFilter("remove")
            }, `Remove VR (${audit.needs_remove.length})`)
          ),

          visible.length === 0 && ce("div", { className: "vra-status" }, "Nothing to show in this category. ✓"),

          ce("div", { className: "vra-list" },
            visible.map(entry => ce("div", {
              key: entry.id,
              className: `vra-row ${doneIds[entry.id] ? "done" : ""}`
            },
              entry.screenshot
                ? ce("img", { src: entry.screenshot, alt: "" })
                : ce("div", { className: "vra-noimg" }),
              ce("div", { className: "vra-row-info" },
                ce("a", {
                  href: `/scenes/${entry.id}`,
                  target: "_blank",
                  className: "vra-row-title"
                }, entry.title),
                ce("div", { className: "vra-row-meta" },
                  `${entry.width}×${entry.height}`,
                  ce("span", { className: "vra-ratio" }, ` (${entry.ratio.toFixed(2)}:1)`)
                )
              ),
              ce("span", {
                className: `vra-action-badge ${entry.action}`
              }, entry.action === "add" ? "+ ADD VR" : "− REMOVE VR"),
              ce("button", {
                className: "vra-btn vra-btn-small",
                disabled: !audit.tag_id || busyIds[entry.id] || doneIds[entry.id],
                onClick: () => applyOne(entry)
              }, doneIds[entry.id] ? "✓ Done" : busyIds[entry.id] ? "…" : "Apply")
            ))
          ),

          ce("div", { className: "vra-actions" },
            visible.length > 0 && ce("button", {
              className: "vra-btn vra-btn-primary",
              disabled: !audit.tag_id,
              onClick: () => applyAll(visible)
            }, `Apply All (${visible.length})`),
            ce("button", {
              className: "vra-btn vra-btn-secondary",
              onClick: reRun, disabled: running
            }, running ? "Re-running…" : "Re-run Audit"),
            ce("button", { className: "vra-btn vra-btn-secondary", onClick: onClose }, "Close")
          )
        )
      )
    );
  }

  // ── Navbar icon ────────────────────────────────────────────────────────────

  async function injectNavButton() {
  const navbar = document.querySelector(".navbar") || document.querySelector("nav");
  if (!navbar) return;

  const target =
    navbar.querySelector(".navbar-buttons") ||
    navbar.querySelector(".ml-auto.navbar-nav") ||
    navbar.querySelector(".navbar-nav:last-child") ||
    navbar;

  // If our button is already inside THIS target, we're done.
  // (getElementById alone isn't enough — the old button may be orphaned
  //  in a detached node after a React re-render.)
  if (target.querySelector("#vra-nav-btn")) return;

  // Clean up any orphaned/stale instances elsewhere in the document
  document.querySelectorAll("#vra-nav-btn").forEach(el => el.remove());

  // Only show the icon if there's an audit available
  const audit = await fetchAudit();
  if (!audit) return;

  const total = audit.needs_add.length + audit.needs_remove.length;

  const btn = document.createElement("button");
  btn.id = "vra-nav-btn";
  btn.title = `VR Audit: ${total} scene${total !== 1 ? "s" : ""} need attention`;
  btn.innerHTML = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="22" height="22" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
    <rect x="2" y="7" width="20" height="11" rx="3"></rect>
    <circle cx="8" cy="12.5" r="2"></circle>
    <circle cx="16" cy="12.5" r="2"></circle>
  </svg>${total > 0 ? `<span class="vra-badge-count">${total}</span>` : ""}`;
  btn.style.cssText = [
    "background:transparent","border:none","color:#aaa","cursor:pointer",
    "padding:6px","display:inline-flex","align-items:center",
    "justify-content:center","border-radius:4px","line-height:1",
    "position:relative",
  ].join(";");
  btn.addEventListener("click", openModal);
  btn.addEventListener("mouseenter", () => { btn.style.color = "#4fc3f7"; });
  btn.addEventListener("mouseleave", () => { btn.style.color = "#aaa"; });

  // prepend() is the safe modern equivalent of insertBefore(x, firstChild)
  // and won't throw if the target's children change underneath us.
  try {
    target.prepend(btn);
    LOG("Nav button injected into:", target.className);
  } catch (e) {
    LOG("Nav button inject failed:", e);
  }
}


)();

File: plugins/MRStashVRTagger/MRStashVRTagger.py
import sys
import json
import os
import urllib.request
import urllib.error
import time

PLUGIN_DIR = os.path.dirname(os.path.abspath(sys.argv[0]))
ASSETS_DIR = os.path.join(PLUGIN_DIR, "assets")
SESSION_COOKIE = None


def graphql_query(url, apikey, query, variables=None):
    headers = {"Content-Type": "application/json"}
    if apikey:
        headers["ApiKey"] = apikey
    elif SESSION_COOKIE:
        headers["Cookie"] = SESSION_COOKIE
    data = json.dumps({"query": query, "variables": variables or {}}).encode("utf-8")
    req = urllib.request.Request(url, data=data, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req) as response:
            body = json.loads(response.read().decode("utf-8"))
            if "errors" in body:
                print(f"GraphQL errors: {body['errors']}", flush=True)
            return body
    except urllib.error.URLError as e:
        print(f"GraphQL request failed: {e}", flush=True)
        sys.exit(1)


def find_or_create_vr_tag(url, apikey, tag_name):
    res = graphql_query(url, apikey, """
    query FindTags($name: String!) {
      findTags(filter: {q: $name}) { tags { id name } }
    }
    """, {"name": tag_name})
    tags = res.get("data", {}).get("findTags", {}).get("tags", [])
    for t in tags:
        if t["name"].lower() == tag_name.lower():
            return t["id"]
    return None


def fetch_all_scenes(url, apikey):
    """Paginate through every scene."""
    all_scenes = []
    page = 1
    per_page = 200
    while True:
        res = graphql_query(url, apikey, """
        query AllScenes($page: Int!, $per_page: Int!) {
          findScenes(filter: {page: $page, per_page: $per_page}) {
            count
            scenes {
              id
              title
              paths { screenshot }
              tags { id name }
              files { width height duration }
            }
          }
        }
        """, {"page": page, "per_page": per_page})
        data = res.get("data", {}).get("findScenes", {})
        scenes = data.get("scenes", [])
        if not scenes:
            break
        all_scenes.extend(scenes)
        total = data.get("count", 0)
        print(f"Fetched {len(all_scenes)}/{total} scenes…", flush=True)
        if len(all_scenes) >= total:
            break
        page += 1
    return all_scenes


def task_audit_vr_tags(url, apikey, plugins_config):
    cfg = plugins_config.get("MRStashVRTagger", {}) or {}
    tag_name = (cfg.get("vr_tag_name") or "Virtual Reality").strip()
    try:
        threshold = float(cfg.get("vr_aspect_ratio") or "1.9")
    except ValueError:
        threshold = 1.9

    print(f"VR tag: {tag_name!r}  threshold: {threshold}", flush=True)

    vr_tag_id = find_or_create_vr_tag(url, apikey, tag_name)
    if not vr_tag_id:
        print(f"Warning: Tag {tag_name!r} not found in library. "
              f"Scenes flagged for ADD won't have a tag to apply until you create it.", flush=True)

    scenes = fetch_all_scenes(url, apikey)
    print(f"Auditing {len(scenes)} scenes…", flush=True)

    needs_add = []      # wide enough, but no VR tag
    needs_remove = []   # has VR tag, but not wide enough
    skipped = []        # no file / no dimensions

    for scene in scenes:
        files = scene.get("files") or []
        if not files:
            skipped.append({"id": scene["id"], "title": scene.get("title"), "reason": "no files"})
            continue
        f = files[0]
        w, h = f.get("width") or 0, f.get("height") or 0
        if not w or not h:
            skipped.append({"id": scene["id"], "title": scene.get("title"), "reason": "no dimensions"})
            continue

        ratio = w / h
        is_wide = ratio >= threshold
        tags = scene.get("tags") or []
        has_vr = any(
            (vr_tag_id and t["id"] == vr_tag_id) or t["name"].lower() == tag_name.lower()
            for t in tags
        )

        entry = {
            "id": scene["id"],
            "title": scene.get("title") or f"Scene #{scene['id']}",
            "screenshot": (scene.get("paths") or {}).get("screenshot"),
            "width": w,
            "height": h,
            "ratio": round(ratio, 3),
        }

        if is_wide and not has_vr:
            entry["action"] = "add"
            needs_add.append(entry)
        elif has_vr and not is_wide:
            entry["action"] = "remove"
            needs_remove.append(entry)

    result = {
        "status": "done",
        "generated_at": int(time.time()),
        "tag_name": tag_name,
        "tag_id": vr_tag_id,
        "threshold": threshold,
        "total_scenes": len(scenes),
        "needs_add": needs_add,
        "needs_remove": needs_remove,
        "skipped_count": len(skipped),
    }

    os.makedirs(ASSETS_DIR, exist_ok=True)
    out_path = os.path.join(ASSETS_DIR, "audit.json")
    with open(out_path, "w") as fp:
        json.dump(result, fp)

    print(f"Done. {len(needs_add)} need ADD, {len(needs_remove)} need REMOVE, "
          f"{len(skipped)} skipped. Wrote {out_path}", flush=True)


def main():
    raw = sys.stdin.read()
    if not raw.strip():
        print("ERROR: stdin is empty.", flush=True)
        sys.exit(1)
    input_data = json.loads(raw)
    sc = input_data.get("server_connection", {})
    scheme = sc.get("Scheme", "http")
    port = sc.get("Port", 9999)
    apikey = sc.get("ApiKey", "")

    global SESSION_COOKIE, PLUGIN_DIR, ASSETS_DIR
    if not apikey:
        cookie = sc.get("SessionCookie", {}) or {}
        if cookie.get("Value"):
            SESSION_COOKIE = f"{cookie.get('Name', 'session')}={cookie['Value']}"

    plugin_dir = sc.get("PluginDir", "")
    if plugin_dir:
        PLUGIN_DIR = plugin_dir
        ASSETS_DIR = os.path.join(PLUGIN_DIR, "assets")
    os.makedirs(ASSETS_DIR, exist_ok=True)

    url = f"{scheme}://localhost:{port}/graphql"

    raw_args = input_data.get("args", {}) or {}
    task_name = raw_args.get("mode", "") if isinstance(raw_args, dict) else ""
    print(f"Task={task_name!r}", flush=True)

    if task_name == "Audit VR Tags":
        config_res = graphql_query(url, apikey, "query { configuration { plugins } }")
        plugins_config = config_res.get("data", {}).get("configuration", {}).get("plugins", {}) or {}
        task_audit_vr_tags(url, apikey, plugins_config)
    else:
        print(f"Unknown task: {task_name!r}", flush=True)
        sys.exit(1)


if __name__ == "__main__":
    main()